### PR TITLE
Add type filtering feature

### LINKED ISSUE #275: How to avoid deserialization of untrusted data attacks with Hyperion serializer?
### Discussed in https://github.com/akkadotnet/akka.net/discussions/5329

<div type='discussions-op-text'>

<sup>Originally posted by **jonnydee** October 19, 2021</sup>
We are currently using Akka 1.4 with Hyperion serialization configured for all user messages sent between Akka systems. We want to make sure Hyperion cannot be tricked into deserialization of untrusted data (see [CWE-502](https://cwe.mitre.org/data/definitions/502.html) for details about that matter).

I know about Hyperion's `akka.actor.serialization-settings.hyperion.disallow-unsafe-type` configuration option which can be used to enable a built-in blacklist for dangerous types. This approach doesn't suffice in our case, however, because we need to trust that the built-in blacklist currently is complete and also will remain complete in the future. Note that being forced to upgrade Akka to some newer release just because new/missing types might have been added to the blacklist is not an option for us. Being able to configure types on the blacklist ourselves would surely be an improvement but the general problem of an outdated blacklist would still remain. It would just be a shift in who is responsible / trusted to have the most recent and complete list of types on the blacklist.

We would like to have an option for a whitelist approach where we list all types that are allowed to be deserialized explicitly. I looked into Hyperion's "Pre-Configured Types" mechanism which involves an [`IKnownTypesProvider`](https://getakka.net/api/Akka.Serialization.IKnownTypesProvider.html). I hoped it would allow us to configure known types explicitly while at the same time ruling out unknown ones. Please correct if I am wrong, but it seems like this is not the case. Instead, known types would be replaced by some id (which is good) but "unknown" types would still be serialized with type information (which is bad). Furthermore, Hyperion's current implementation for mapping ids to known types seems to depend on the order of the types returned by `IKnownTypesProvider.GetKnownTypes()` method. (The id corresponds to the type's position in the returned `IEnumerable<Type>`.) This makes an approach where known message types are collected by some automatism (e.g. a reflection-based approach) difficult, because all involved Actor Systems would have to see exactly the same list (size/items/order) of known types even though some of them would never see all messages in that "overall" list.

We know that using some schema-based serialization technology (e.g. Google's Protocol Buffers) doesn't suffer from such attack vectors but, unfortunately, we cannot use it in our product. Also securing communication between Actor Systems using a VPN is not an option for us.

So the questions are:

- Is there some other way to defend against untrusted data attacks when using Hyperion?
- Do you have any idea on how to implement a whiltelisting machanism for Hyperion in an easy and future-proof way?</div>

We need to find a way to configure black and white listing of types for the Hyperion serializer.

- [x] Add Whitelist/blacklist implementation
- [ ] Improve KnownValueProvider to not depend on array index for its manifest id

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #275

## Changes

Implemented the whitelist/blacklist spec outlined in #275

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #275
* [x] Changes in public API reviewed, if any.
* [x] I have added website documentation for this feature.

### CODE REVIEW

File: src/Hyperion/Extensions/TypeEx.cs
Comment (Arkatufus) [inline src/Hyperion/Extensions/TypeEx.cs:202]: The filter gets called here.

File: src/Hyperion/SerializerOptions.cs
Comment (Arkatufus) [inline src/Hyperion/SerializerOptions.cs:123]: ITypeFilter is passed into the options here

File: src/Hyperion/TypeFilter.cs
Comment (Arkatufus) [inline src/Hyperion/TypeFilter.cs:9]: ITypeFilter implementation

File: src/Hyperion/TypeFilter.cs
Comment (Arkatufus) [inline src/Hyperion/TypeFilter.cs:22]: Default ITypeFilter that always returns true, used when type filtering is disabled.

File: src/Hyperion/TypeFilterBuilder.cs
Comment (Arkatufus) [inline src/Hyperion/TypeFilterBuilder.cs:6]: Simple builder pattern for creating a TypeFilter instance

File: src/Hyperion.API.Tests/CoreApiSpec.ApproveApi.approved.txt
Comment (eaba) [inline src/Hyperion.API.Tests/CoreApiSpec.ApproveApi.approved.txt:55]: LGTM

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Arkatufus) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:68]: Yes, this is a security feature on the receiving end, we're trying to harden the deserializer from possible poisoned injection attack.

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Arkatufus) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:38]: We would have to eventually, but this can't be tested against Akka.NET from Hyperion side until we do a release of Hyperion since Akka.NET is dependent on Hyperion, not the other way around.
It would make more sense to test this on the Akka.NET side instead of on the Hyperion side.

File: src/Hyperion/Extensions/TypeEx.cs
Comment (eaba) [inline src/Hyperion/Extensions/TypeEx.cs:200]: How do we safely check if a type is not custom type designed to bypass this guard? Also, if `System.Security.Claims.ClaimsIdentity` is contained in `UnsafeTypesDenySet` `name.Contains(Claims.ClaimsIdentity)` will will return true, right?

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (eaba) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:38]: Based on my discussion with @Aaronontheweb, we could create an integration test on the Hyperion test suite so that changes are thoroughly tested against Akka.NET before Nuget package is released. The benefit is that when we fail to capture it in code review, the test will at least assist with that.

File: src/Hyperion/Extensions/TypeEx.cs
Comment (eaba) [reply src/Hyperion/Extensions/TypeEx.cs:200]: Ok

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Aaronontheweb) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:38]: > to eventually, but this can't be tested against Akka.NET from Hyperion side until we do a release of Hyperion since Akka.NET is dependent on Hyperion, not the other way around.
> It would make more sense to test this on the Akka.NET side instead of on the Hyperion

What @eaba is asking is that this test doesn't cover the same `Deserialze<object>` routine we use in Akka.NET itself and that could be added to this test suite here. We will have to do a separate PR for Akka.NET's consumption of this feature, of course, but we should also test that the way Akka.NET uses Hyperion is supported by this feature prior to releasing this package. Does that make sense @Arkatufus ?

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Arkatufus) [inline src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:59]: Would this be enough to satistfy the Akka.NET test requirement?

Comment (Aaronontheweb) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:59]: Absolutely

Comment (Aaronontheweb) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:59]: Should also have a version of that that throws too when bad input is sent

Comment (Arkatufus) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:59]: There is, its at the end of the test

**Comment by Aaronontheweb** [comment]
@Arkatufus looks like API approval failed

**Comment by Aaronontheweb** [comment]
One final thing we need before approval - a before / after benchmark on Hyperion. You'll need to do three benchmarks:

1. [ ] Current `dev` branch throughput
2. [ ] This PR throughput, with the TypeFilter unpopulated
3. [ ] This PR throughput, with the TypeFilter populated

Need to measure the impact of this, if any, before we start recommending it to users.

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Aaronontheweb) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:38]: @Arkatufus handled this further down the PR

File: src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs
Comment (Aaronontheweb) [reply src/Hyperion.Tests/UnsafeDeserializationExclusionTests.cs:59]: LGTM

**Comment by Arkatufus** [comment]
Before changes, current dev branch benchmark:

``` ini
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19041.1415 (2004/May2020Update/20H1)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4420.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4420.0), X64 RyuJIT
```

#### Hyperion.Benchmarks.SerializeClassesBenchmark

|               Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|--------------------- |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|    Cyclic_References |   951.2 ns |  5.71 ns |  5.34 ns |   941.6 ns |   957.8 ns | 1,051,353.3 | 0.1888 |      1 KB |
|      Virtual_Classes |   726.5 ns |  3.02 ns |  2.67 ns |   721.6 ns |   729.5 ns | 1,376,511.5 | 0.1850 |      1 KB |
| Large_Sealed_Classes | 3,237.5 ns | 13.24 ns | 11.74 ns | 3,223.6 ns | 3,255.9 ns |   308,878.6 | 0.4311 |      3 KB |
|      Generic_Classes | 1,861.7 ns |  9.59 ns |  8.01 ns | 1,851.1 ns | 1,881.5 ns |   537,150.1 | 0.2174 |      1 KB |

#### Hyperion.Benchmarks.SerializeStructsBenchmark

|                Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|---------------------- |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|                 Enums |   360.6 ns |  6.94 ns |  6.49 ns |   350.8 ns |   371.1 ns | 2,773,330.7 | 0.0925 |     586 B |
|  Standard_Value_Types | 1,136.9 ns | 17.59 ns | 15.59 ns | 1,124.0 ns | 1,171.5 ns |   879,557.1 | 0.1163 |     738 B |
| Blittable_Value_Types |   707.3 ns | 13.66 ns | 14.03 ns |   695.0 ns |   732.2 ns | 1,413,892.3 | 0.1001 |     634 B |

**Comment by Arkatufus** [comment]
After changes, with filtering turned on and off:

### .NET Framework 4.7.1

#### Hyperion.Benchmarks.SerializeClassesBenchmark

|                        Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|------------------------------ |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|             Cyclic_References |   953.7 ns | 17.47 ns | 16.34 ns |   933.9 ns |   983.0 ns | 1,048,575.3 | 0.1888 |      1 KB |
|    Filtered_Cyclic_References |   948.1 ns | 13.29 ns | 12.43 ns |   934.6 ns |   967.5 ns | 1,054,725.3 | 0.1888 |      1 KB |
|               Virtual_Classes |   742.0 ns |  9.22 ns |  8.63 ns |   730.7 ns |   755.2 ns | 1,347,630.2 | 0.1850 |      1 KB |
|      Filtered_Virtual_Classes |   726.8 ns |  5.20 ns |  4.61 ns |   717.4 ns |   733.2 ns | 1,375,876.2 | 0.1850 |      1 KB |
|          Large_Sealed_Classes | 3,286.2 ns | 42.76 ns | 39.99 ns | 3,245.1 ns | 3,373.8 ns |   304,301.0 | 0.4311 |      3 KB |
| Filtered_Large_Sealed_Classes | 3,397.2 ns | 37.53 ns | 35.10 ns | 3,356.0 ns | 3,471.8 ns |   294,359.3 | 0.4349 |      3 KB |
|               Generic_Classes | 1,894.1 ns | 22.91 ns | 21.43 ns | 1,872.2 ns | 1,933.9 ns |   527,957.2 | 0.2174 |      1 KB |
|      Filtered_Generic_Classes | 1,882.8 ns | 15.09 ns | 13.38 ns | 1,858.9 ns | 1,909.2 ns |   531,119.2 | 0.2174 |      1 KB |

#### Hyperion.Benchmarks.SerializeStructsBenchmark

|                         Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|------------------------------- |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|                          Enums |   363.2 ns |  4.49 ns |  4.20 ns |   358.6 ns |   371.7 ns | 2,753,108.5 | 0.0925 |     586 B |
|                 Filtered_Enums |   354.5 ns |  1.83 ns |  1.62 ns |   352.4 ns |   357.8 ns | 2,821,200.8 | 0.0925 |     586 B |
|           Standard_Value_Types | 1,139.5 ns | 18.75 ns | 17.54 ns | 1,116.7 ns | 1,176.0 ns |   877,557.5 | 0.1163 |     738 B |
|  Filtered_Standard_Value_Types | 1,122.3 ns | 16.51 ns | 14.64 ns | 1,105.5 ns | 1,159.4 ns |   891,021.4 | 0.1163 |     738 B |
|          Blittable_Value_Types |   699.1 ns |  4.89 ns |  4.33 ns |   693.4 ns |   710.0 ns | 1,430,464.6 | 0.1001 |     634 B |
| Filtered_Blittable_Value_Types |   694.6 ns |  2.86 ns |  2.39 ns |   689.8 ns |   699.1 ns | 1,439,615.2 | 0.1001 |     634 B |

### .NET Core 3.1

#### Hyperion.Benchmarks.SerializeClassesBenchmark

|                        Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|------------------------------ |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|             Cyclic_References | 1,039.5 ns |  9.32 ns |  7.79 ns | 1,029.0 ns | 1,057.7 ns |   961,957.4 | 0.1373 |      1 KB |
|    Filtered_Cyclic_References | 1,050.5 ns | 10.98 ns | 10.27 ns | 1,030.3 ns | 1,063.0 ns |   951,943.0 | 0.1373 |      1 KB |
|               Virtual_Classes |   811.9 ns |  4.46 ns |  3.73 ns |   804.4 ns |   818.3 ns | 1,231,608.3 | 0.1354 |      1 KB |
|      Filtered_Virtual_Classes |   845.3 ns | 15.41 ns | 14.42 ns |   827.0 ns |   876.3 ns | 1,183,016.7 | 0.1354 |      1 KB |
|          Large_Sealed_Classes | 3,216.7 ns | 64.12 ns | 73.84 ns | 3,129.1 ns | 3,359.6 ns |   310,881.0 | 0.3166 |      3 KB |
| Filtered_Large_Sealed_Classes | 3,029.8 ns | 52.66 ns | 49.26 ns | 2,973.1 ns | 3,114.0 ns |   330,051.3 | 0.3128 |      3 KB |
|               Generic_Classes | 1,766.7 ns |  8.81 ns |  8.24 ns | 1,753.5 ns | 1,784.7 ns |   566,019.0 | 0.1678 |      1 KB |
|      Filtered_Generic_Classes | 1,743.6 ns | 21.89 ns | 18.28 ns | 1,724.5 ns | 1,786.0 ns |   573,524.0 | 0.1678 |      1 KB |

#### Hyperion.Benchmarks.SerializeStructsBenchmark

|                         Method |     Mean |    Error |   StdDev |      Min |        Max |        Op/s |  Gen 0 | Allocated |
|------------------------------- |---------:|---------:|---------:|---------:|-----------:|------------:|-------:|----------:|
|                          Enums | 353.9 ns |  2.61 ns |  2.44 ns | 348.3 ns |   356.6 ns | 2,825,836.0 | 0.0677 |     568 B |
|                 Filtered_Enums | 355.2 ns |  3.25 ns |  2.53 ns | 352.3 ns |   361.8 ns | 2,815,647.7 | 0.0677 |     568 B |
|           Standard_Value_Types | 973.3 ns |  7.50 ns |  6.65 ns | 963.1 ns |   988.7 ns | 1,027,414.0 | 0.0839 |     712 B |
|  Filtered_Standard_Value_Types | 992.9 ns | 19.26 ns | 18.91 ns | 968.5 ns | 1,038.2 ns | 1,007,200.1 | 0.0839 |     712 B |
|          Blittable_Value_Types | 581.3 ns |  5.65 ns |  5.01 ns | 576.2 ns |   590.4 ns | 1,720,428.4 | 0.0734 |     616 B |
| Filtered_Blittable_Value_Types | 598.0 ns | 10.29 ns |  9.62 ns | 587.0 ns |   616.7 ns | 1,672,306.3 | 0.0734 |     616 B |

### .NET 5.0

#### Hyperion.Benchmarks.SerializeClassesBenchmark

|                        Method |       Mean |    Error |   StdDev |        Min |        Max |        Op/s |  Gen 0 | Allocated |
|------------------------------ |-----------:|---------:|---------:|-----------:|-----------:|------------:|-------:|----------:|
|             Cyclic_References | 1,031.6 ns |  7.56 ns |  7.07 ns | 1,023.3 ns | 1,046.4 ns |   969,401.0 | 0.1392 |      1 KB |
|    Filtered_Cyclic_References | 1,025.6 ns | 19.94 ns | 21.33 ns |   999.1 ns | 1,069.1 ns |   975,035.1 | 0.1392 |      1 KB |
|               Virtual_Classes |   792.8 ns |  8.13 ns |  7.21 ns |   785.4 ns |   810.6 ns | 1,261,383.9 | 0.1364 |      1 KB |
|      Filtered_Virtual_Classes |   797.1 ns | 11.66 ns | 10.33 ns |   777.8 ns |   816.5 ns | 1,254,604.7 | 0.1364 |      1 KB |
|          Large_Sealed_Classes | 3,097.2 ns | 58.58 ns | 57.53 ns | 3,028.3 ns | 3,189.6 ns |   322,873.5 | 0.3166 |      3 KB |
| Filtered_Large_Sealed_Classes | 3,054.4 ns | 41.82 ns | 39.12 ns | 2,998.0 ns | 3,130.9 ns |   327,399.5 | 0.3166 |      3 KB |
|               Generic_Classes | 1,716.6 ns |  6.53 ns |  6.11 ns | 1,708.4 ns | 1,727.6 ns |   582,541.2 | 0.1698 |      1 KB |
|      Filtered_Generic_Classes | 1,668.2 ns | 16.77 ns | 14.87 ns | 1,651.6 ns | 1,705.6 ns |   599,431.8 | 0.1698 |      1 KB |

#### Hyperion.Benchmarks.SerializeStructsBenchmark

|                         Method |     Mean |   Error |  StdDev |      Min |      Max |        Op/s |  Gen 0 | Allocated |
|------------------------------- |---------:|--------:|--------:|---------:|---------:|------------:|-------:|----------:|
|                          Enums | 349.4 ns | 4.22 ns | 3.95 ns | 343.8 ns | 355.7 ns | 2,861,845.7 | 0.0677 |     568 B |
|                 Filtered_Enums | 344.6 ns | 2.89 ns | 2.42 ns | 340.2 ns | 349.1 ns | 2,901,886.7 | 0.0677 |     568 B |
|           Standard_Value_Types | 938.1 ns | 5.36 ns | 4.48 ns | 932.9 ns | 946.1 ns | 1,066,020.2 | 0.0849 |     712 B |
|  Filtered_Standard_Value_Types | 937.6 ns | 5.68 ns | 5.32 ns | 928.8 ns | 947.9 ns | 1,066,550.0 | 0.0849 |     712 B |
|          Blittable_Value_Types | 575.1 ns | 9.19 ns | 8.60 ns | 562.4 ns | 584.8 ns | 1,738,796.2 | 0.0734 |     616 B |
| Filtered_Blittable_Value_Types | 586.3 ns | 7.48 ns | 6.25 ns | 572.9 ns | 597.5 ns | 1,705,694.0 | 0.0734 |     616 B |

**Review by Aaronontheweb — APPROVED** [review]
Changes look good, benchmark is within the margin of error in terms of total performance impact. :shipit:

**Pull request merged**
